### PR TITLE
Dynamically link NCCL EP for USE_SYSTEM_NCCL=ON (wheel) builds

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #187385

Allow `USE_NCCL_EP` with `USE_SYSTEM_NCCL=ON` (drop the `NOT USE_SYSTEM_NCCL` gate) and branch the
extension's libnccl_ep linkage to **follow torch's NCCL linkage**:

- `USE_SYSTEM_NCCL=OFF` (source / dev build): static libnccl_ep.a, self-contained (as the previous commit).
- `USE_SYSTEM_NCCL=ON` (wheel / CI): torch dynamically links the system NCCL, so libnccl_ep.so is built
  against that same NCCL and the extension NEEDED-links it, resolved from the nccl4py wheel at runtime
  (JIT headers from nccl4py). 

Review order:
1. CMakeLists.txt: drop the `NOT USE_SYSTEM_NCCL` gate for NCCL EP.
2. cmake/External/nccl_ep.cmake: branch shared `.so` (system NCCL) vs static `.a`; bake `NCCL_EP_JIT_HOME`
   only on the static path.
3. torch/CMakeLists.txt: dynamic path NEEDED-links libnccl_ep + uses the nccl4py RPATH.
4. `_token_switch.py`: `_prepare_nccl4py` locate nccl4py, set `NCCL_EP_HOME/NCCL_HOME`, load
   libnccl_ep.

Test Plan:
`USE_SYSTEM_NCCL=ON` on 4xH100, using the nvidia.nccl wheel as the system NCCL: the extension NEEDED-links
libnccl_ep.so.0, libtorch_cuda NEEDED-links libnccl.so.2, and the suite passes through the
_prepare_nccl4py fallback (Ran 7 tests ... OK).

### CODE REVIEW

File: torch/CMakeLists.txt [hunk 1]
```diff
@@ -472,27 +472,38 @@ if(BUILD_PYTHON AND NOT BUILD_LIBTORCH_WHL)
...
       torch_python torch_cuda __caffe2_nccl_ep pybind::pybind11)
     set_target_properties(_nccl_ep PROPERTIES
-      LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/torch"
-      INSTALL_RPATH "${_rpath_portable_origin}/lib")
+      LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/torch")
+    if(USE_SYSTEM_NCCL)
+      # Dynamic: resolve nccl4py's libnccl_ep.so (site-packages/nccl/ep/lib, one
+      # dir up from torch/ via $ORIGIN/..) at import.
+      set_target_properties(_nccl_ep PROPERTIES
+        INSTALL_RPATH "${_rpath_portable_origin}/lib:${_rpath_portable_origin}/../nccl/ep/lib")
```
[the remark this answers is not in the document]
Comment (kwen2501) [reply torch/CMakeLists.txt]: Done -- switched to a `;`-separated CMake list (`${_rpath_portable_origin}/lib;${_rpath_portable_origin}/../nccl/ep/lib`), which is the idiomatic form.

File: torch/distributed/_token_switch.py [hunk 2]
```diff
@@ -12,18 +15,81 @@
...
+    if nccl_pkg is None:
+        raise ImportError(
+            "TokenSwitchNCCL needs the 'nccl4py' package for this build's "
+            "libnccl_ep.so and runtime JIT headers. Install it with "
+            "`pip install nccl4py==0.3.1`."
+        )
+    ep_dir = os.path.join(nccl_pkg, "ep")
+    if os.path.isdir(os.path.join(ep_dir, "include", "nccl_ep")):
+        os.environ.setdefault("NCCL_EP_HOME", ep_dir)
+
```
[the remark this answers is not in the document]
Comment (kwen2501) [reply torch/distributed/_token_switch.py:58]: Done. `_prepare_nccl4py` now raises an `ImportError` naming the expected `<nccl4py>/ep/include/nccl_ep` path when it has to set `NCCL_EP_HOME` but those headers are absent, instead of skipping silently. It still defers to a user-set `NCCL_EP_HOME`.

File: torch/distributed/_token_switch.py [hunk 3]
```diff
@@ -12,18 +15,81 @@
...
+    nvidia_nccl = _find_pkg_dir("nvidia.nccl")
+    if nvidia_nccl is not None and os.path.isdir(os.path.join(nvidia_nccl, "include")):
+        os.environ.setdefault("NCCL_HOME", nvidia_nccl)
+
+    # nccl4py ships libnccl_ep unversioned (libnccl_ep.so) while the extension's
+    # NEEDED entry is its SONAME libnccl_ep.so.0; load it by path to register
+    # that SONAME for the import below.
+    lib = os.path.join(ep_dir, "lib", "libnccl_ep.so")
+    if os.path.isfile(lib):
+        _NCCL_EP_KEEPALIVE.append(ctypes.CDLL(lib, mode=ctypes.RTLD_LOCAL))
```
[the remark this answers is not in the document]
Comment (kwen2501) [reply torch/distributed/_token_switch.py:72]: Leaving this as best-effort rather than a hard raise, on purpose. The preload is a workaround for nccl4py shipping the *unversioned* `libnccl_ep.so` while the extension's NEEDED entry is the SONAME `libnccl_ep.so.0`. If nccl4py is fixed to ship `libnccl_ep.so.0` (an issue we've filed), the import resolves directly via RPATH with no preload needed -- a hard raise on the missing unversioned file would then wrongly reject a perfectly working install. The import itself is the authoritative check, and the clearer header error (other thread) covers the common misconfiguration.

File: torch/distributed/_token_switch.py [hunk 4]
```diff
@@ -12,18 +15,81 @@
...
+    # nccl4py. A USE_SYSTEM_NCCL=ON build NEEDED-links libnccl_ep from the nccl4py
+    # wheel, so the first import fails until nccl4py is set up; fall back to that
+    # and retry.
+    try:
+        # pyrefly: ignore [missing-import]  # built only with USE_NCCL_EP
+        import torch._nccl_ep as _ep
+
+        return _ep
+    except ImportError:
+        pass
```
[the remark this answers is not in the document]
Comment (kwen2501) [reply torch/distributed/_token_switch.py:88]: Leaving the fallback as-is. In the common case (nccl4py present) the first and second import fail identically, and the retry's `raise ... from e` already chains the real error, so it isn't masked. The masking window is narrow -- an import failure unrelated to nccl4py *and* nccl4py absent -- and distinguishing it by matching on the error string is fragile, so I'd rather not add that coupling.

**Comment by kwen2501** [comment]
@Skylion007 Thanks! I've resolved the comments from Copilot.

**Review by kapilsh — APPROVED** [review]
Looks good!

File: torch/distributed/_token_switch.py [hunk 5]
```diff
@@ -12,18 +15,87 @@
     from torch.distributed.distributed_c10d import ProcessGroup
 
 
+# Keeps preloaded shared libraries resident for the process lifetime.
+_NCCL_EP_KEEPALIVE: list[ctypes.CDLL] = []
+
+
+def _find_pkg_dir(name: str) -> str | None:
```
Comment (Skylion007) [inline torch/distributed/_token_switch.py:22]: Do a search, I think we have a shared utility for this somewhere, and if not we should.

File: torch/distributed/_token_switch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Skylion007) [reply torch/distributed/_token_switch.py:88]: Can we call logger.debug here with exc_info=True then?

**Review by Skylion007 — APPROVED** [review]
